content/blog/2025-09-15-hash-fields-expiration/index.md

```diff
@@ -0,0 +1,225 @@
+2. The [`HRANDFIELD`](/commands/hrandfield/) command selects fields at random from the underlying hash object.
+   If a hash contains many expired fields awaiting deletion, the selection process may frequently land on those expired entries.
+   In practice, this can result in [`HRANDFIELD`](/commands/hrandfield/) returning an empty reply more often than expected, and it reduces the fairness of field sampling.
+3. Valkey normally uses a highly compact representation for small hashes, which saves significant memory.
+   To simplify the expiration design in Valkey 9.0, we chose to disable this compact encoding when a hash contains volatile fields (i.e., fields with TTLs).
+   The impact is limited: only small hashes with field expirations are affected, but in those cases memory usage will be slightly higher than the fully compacted representation.
+
+## Benchmarking our solution
+
+Validating the new design meant benchmarking across several dimensions: memory overhead, command performance, and expiration efficiency.
+
+### Memory Overhead
+
+We first measured the per-field memory overhead when setting TTLs.
+The raw expiration time itself requires a constant 8 bytes (though this could be reduced in the future
+by storing only a delta relative to a reference timestamp, such as server start time).
+On top of that, extra memory is needed for tracking within the vset.
+
+The actual overhead depends on both how many fields have expirations and how spread out their expiration times are.
+This is because the bucket encoding chosen by the vset adapts to the data distribution.
+In practice, the overhead ranged between 16 and 29 bytes per field.
+
+![illustration](hfe-benchmark-memory.png)
+
+The higher end of this range primarily affects small hashes, where compact encodings
+like listpack are avoided when volatile fields are present.
+
+### Command Performance
+
+Next, we benchmarked common hash commands both with and without field expirations.
+The results showed no measurable performance regression and throughput remained stable when expirations were added.
+
+![illustration](hfe-benchmark-hash-commands.png)
+
+We also benchmarked the new expiration-aware commands (e.g., `HSETEX`), confirming that their performance is on par with traditional hash operations.
+
+![illustration](hfe-benchmark-new-commands.png)
+
+### Active Expiration Efficiency
+
+The design goal of the vset was to enable efficient background deletion of expired fields.
+To test this, we preloaded 10 million fields with TTLs.
+We distributed these fields across varying numbers of hash objects to see how object size influences expiration.
+During the load phase, we disabled the expiration job using the [`DEBUG`](/commands/debug/), then re-enabled it once all fields had expired.
+
+The following chart shows the time it took the expiration cron job complete the full deletion of all the 10M fields.
+
+![illustration](hfe-benchmark-new-commands.png)
```

Comment:
I had the question when were discussing this in Amsterdam about what the CPU was at.

Reply:
Let me see if I can add this

Reply:
<img width="601" height="382" alt="image" src="https://github.com/user-attachments/assets/8756de7a-cf91-4675-b3ea-bc6717c610f7" />

This is how it will show + CPU%

Is that better?